crates/typstyle-core/src/partial.rs

```diff
@@ -0,0 +1,126 @@
+use std::ops::Range;
+
+use typst_syntax::{
+    ast::{Expr, Markup},
+    LinkedNode, Source, Span, SyntaxKind,
+};
+
+use crate::{pretty::Mode, utils, AttrStore, FormatError, PrettyPrinter, Typstyle, TypstyleConfig};
+
+impl Typstyle {
+    /// Format the node with minimal span that covering the given *byte* range.
+    /// If you have a utf16 range instead, use [`utf16_range_to_byte_range`] to construct the byte range.
+    pub fn format_byte_range(
+        &self,
+        source: &Source,
+        byte_range: Range<usize>,
+    ) -> Result<(Range<usize>, String), FormatError> {
+        let range = utils::trim_range(source.text(), byte_range);
+
+        let Some((node, mode)) =
+            get_expr_cover_range(source, range.clone()).filter(|(node, _)| !node.erroneous())
+        else {
+            return Err(FormatError::SyntaxError);
+        };
+
+        let printer = PrettyPrinter::new(TypstyleConfig::default(), AttrStore::new(source.root()));
+        printer.push_mode(mode);
+        let doc = if let Some(expr) = node.cast() {
+            printer.convert_expr(expr)
+        } else if let Some(markup) = node.cast() {
+            printer.convert_markup(markup)
+        } else {
+            return Err(FormatError::SyntaxError);
+        };
+        let indent = utils::count_spaces_after_last_newline(source.text(), range.start);
+        let res = doc
+            .nest(indent as isize)
+            .pretty(self.config.max_width)
+            .to_string();
+        Ok((range, res))
+    }
+}
+
+pub fn utf16_range_to_byte_range(
```

Comment:
We don't have to process the utf-16 encoding in typstyle. The other programs like tinymist will preprocess utf-16 things for typstyle.